src/components/views/rooms/UserIdentityWarning.tsx

```diff
@@ -52,277 +49,235 @@ async function userNeedsApproval(crypto: CryptoApi, userId: string): Promise<boo
-        this.state = {
-            currentPrompt: undefined,
-        };
-        this.membersNeedingApproval = new Map();
-        this.gotVerificationStatusUpdate = new Map();
-        this.mounted = true;
-        this.initialised = false;
-        this.addListeners();
-    }
+export const UserIdentityWarning: React.FC<UserIdentityWarningProps> = ({ room }) => {
+    const cli = MatrixClientPeg.safeGet();
+    const crypto = cli.getCrypto();
 
-    public componentDidMount(): void {
-        if (!MatrixClientPeg.safeGet().getCrypto()) return;
-        if (this.props.room.hasEncryptionStateEvent()) {
-            this.initialise().catch((e) => {
-                logger.error("Error initialising UserIdentityWarning:", e);
-            });
-        }
-    }
+    // The current room member that we are prompting the user to approve.
+    const [currentPrompt, setCurrentPrompt] = useState<RoomMember | undefined>(undefined);
 
-    public componentWillUnmount(): void {
-        this.mounted = false;
-        this.removeListeners();
-    }
+    // Whether or not we've already initialised the component by loading the
+    // room membership.
+    const initialisedRef = useRef<boolean>(false);
+    // Which room members need their identity approved.
+    const membersNeedingApprovalRef = useRef<Map<string, RoomMember>>(new Map());
+    // Whether we got a verification status update while we were fetching a
+    // user's verification status.
+    //
+    // We set the entry for a user to `false` when we fetch a user's
+    // verification status, and remove the user's entry when we are done
+    // fetching.  When we receive a verification status update, if the entry for
+    // the user is `false`, we set it to `true`.  After we have finished
+    // fetching the user's verification status, if the entry for the user is
+    // `true`, rather than `false`, we know that we got an update, and so we
+    // discard the value that we fetched.  We always use the value from the
+    // update and consider it as the most up-to-date version.  If the fetched
+    // value is more up-to-date, then we should be getting a new update soon
+    // with the newer value, so it will fix itself in the end.
+    const gotVerificationStatusUpdateRef = useRef<Map<string, boolean>>(new Map());
 
-    /**
-     * Select a new user to display a warning for.  This is called after the
-     * current prompted user no longer needs their identity approved.
-     */
-    private selectCurrentPrompt(): void {
-        if (this.membersNeedingApproval.size === 0) {
-            this.setState({
-                currentPrompt: undefined,
-            });
-            return;
-        }
-        // We return the user with the smallest user ID.
-        const keys = Array.from(this.membersNeedingApproval.keys()).sort((a, b) => a.localeCompare(b));
-        this.setState({
-            currentPrompt: this.membersNeedingApproval.get(keys[0]!),
-        });
-    }
+    useEffect(() => {
```

Comment:
It's difficult to follow the behaviour of the hook. This  hook can be divided into multiple custom hooks, where each one has it's own logic. You can encapsulate the logic of each group and their variable/properties. cf (https://react.dev/learn/reusing-logic-with-custom-hooks)

The fact that we are using 3 refs to store data make me feel that we can improve and simplify the logic here.

To listen to the `MatrixClient` event, you can use the hook `useTypedEventEmitter`.

Reply:
I've split it into multiple hooks now.  I don't really see how to get rid of the 3 refs.